src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -71,105 +87,87 @@ void ucc_tl_ucp_reduce_scatter_knomial_progress(ucc_coll_task_t *coll_task)
+                                   peer, team, task),
+                task, out);
             rbuf = PTR_OFFSET(rbuf, local_seg_count * dt_size);
         }
-    UCC_KN_PHASE_LOOP:
+UCC_KN_PHASE_LOOP:
         if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
             SAVE_STATE(UCC_KN_PHASE_LOOP);
             return;
         }
         if (task->tagged.send_posted > p->iteration * (radix - 1)) {
-            sbuf       = (ucc_knomial_pattern_loop_first_iteration(p))
-                             ? ((KN_NODE_PROXY == node_type || UCC_IS_INPLACE(*args))
-                                    ? args->dst.info.buffer
-                                    : args->src.info.buffer)
-                             : task->reduce_scatter_kn.scratch;
-            rbuf       = (!ucc_knomial_pattern_loop_first_iteration(p))
-                             ? PTR_OFFSET(task->reduce_scatter_kn.scratch,
-                                    block_count * dt_size)
-                             : task->reduce_scatter_kn.scratch;
             step_radix = ucc_kn_compute_step_radix(p);
-            local_seg_index =
-                ucc_kn_compute_seg_index(rank, p->radix_pow, p);
-            local_seg_count = ucc_sra_kn_compute_seg_size(
-                block_count, step_radix, local_seg_index);
-            local_seg_offset = ucc_sra_kn_compute_seg_offset(
-                block_count, step_radix, local_seg_index);
+            ucc_kn_rs_pattern_peer_seg(rank, p, &local_seg_count,
+                                       &local_seg_offset);
+            get_sbuf_rbuf(p, args, task->reduce_scatter_kn.scratch, block_count,
+                          &sbuf, &rbuf);
             local_data  = PTR_OFFSET(sbuf, local_seg_offset * dt_size);
             reduce_data = task->reduce_scatter_kn.scratch;
             is_avg      = args->op == UCC_OP_AVG &&
                      (avg_pre_op ? ucc_knomial_pattern_loop_first_iteration(p)
                                  : ucc_knomial_pattern_loop_last_iteration(p));
+            ucc_assert((step_radix - 1) ==
+                       (task->tagged.send_posted - p->iteration * (radix - 1)));
 
-            if (task->reduce_scatter_kn.scratch_mc_header &&
+            if (!task->reduce_scatter_kn.scratch_mc_header &&
                 ucc_knomial_pattern_loop_last_iteration(p)) {
-                ucc_sra_kn_get_offset_and_seglen(count, dt_size, rank, size, radix,
-                                                 &offset, &local_seg_count);
-                reduce_data = PTR_OFFSET(args->dst.info.buffer, offset);
+                status = ucc_dt_reduce_strided(
+                    rbuf, PTR_OFFSET(rbuf, local_seg_count * dt_size), rbuf,
+                    step_radix - 2, local_seg_count, local_seg_count * dt_size,
+                    dt, args, 0, 0, task->reduce_scatter_kn.executor,
+                    &task->reduce_scatter_kn.etask);
+
+            } else {
+                if (task->reduce_scatter_kn.scratch_mc_header &&
```

Comment:
maybe i missing something, there 3 different cases 
* last iteration and don't have scratch - partial reduction
* last iteration and have scratch  - full reduction since receive data is in scratch
* not last iteration - full reduction